rfcs/005-all-docs-index.md

```diff
@@ -0,0 +1,207 @@
+  on disk.
+* `sizes.file`: the current number of bytes on disk.
+
+The relationship between `sizes.active` and `sizes.file` is used to guide
+decisions on database compaction. FoundationDB doesn't require compaction, and
+any distinction that might exist between these two quantities (e.g. from storage
+engine compression) is not surfaced up to the clients, so it probably doesn't
+make sense to have both.
+
+The current implementation of `sizes.external` does *not* measure the length of
+a JSON representation of the data, but rather the size of an uncompressed Erlang
+term representation of the JSON. This is a somewhat awkward choice as the
+internal Erlang term representation is liable to change over time (e.g. with the
+introduction of Maps in newer Erlang releases, or plausibly even a JSON decoder
+that directly emits the format defined in the document storage RFC).
+
+Assuming we can agree on a set of sizes and how they should be calculated, the
+implementation will require two pieces: a single key for each size, mutated by
+atomic operations, and a record of the size of each revision in the ?REVISIONS
+subpsace so that a transaction can compute the delta for each document.
+
+### Clustering
+
+The `r`, `w`, `q`, and `n` values in the `cluster` object were introduced in
+CouchDB 2.x to describe the topology of a database and the default quorum
+settings for operations against it. If we wanted to bring these forward, here's
+how they'd be defined:
+
+* `r`: always fixed at 1
+
+* `w`: interpreted as the number of transaction logs that record a commit, this
+  is dependent on the `redundancy mode` for the underlying FoundationDB database
+
+* `n`: interpreted as number of storage servers that host a key, this is also
+  dependent on the `redundancy mode` for the underlying FoundationDB database
+
+* `q`: the closest analogue here would be to use the `get_boundary_keys` API and
+  report number of distinct ranges implied by the boundary keys
+
+This interpretation could lead to some surprises, though. For example, "r=1,
+w=4, n=3" is a popular configuration, but this is nonsensical for someone
+expecting to see Dynamo-style numbers. Ignoring backwards compatibility, the
+sensible thing is to point users toward the actual FoundationDB configuration
+information, and to deprecate this entire `cluster` object. Open for discussion.
+
+# Advantages and Disadvantages
+
+[NOTE]: # ( Briefly, list the benefits and drawbacks that would be realized should )
+[NOTE]: # ( the proposal be accepted for inclusion into Apache CouchDB. )
+
+# Key Changes
+
+The underlying transaction in FoundationDB must complete within 5 seconds, which
+implicitly limits the number of results that can be returned in a single
+_all_docs invocation.
+
+## Applications and Modules affected
+
+TBD depending on exact code layout going forward.
+
+## HTTP API additions
+
+None.
+
+## HTTP API deprecations
+
+The `total_rows` and `offset` fields are removed from the response to
```

Comment:
Why are we dropping those and if we do, how do we tell users to “iterate over all docs in a db”? I realise the semantics of this are going to change with the 5s transaction limit, but users need at least guidance on how to do this, and what to expect.

Reply:
My initial thinking was to drop `offset` because there's a fair bit of overhead to compute it (FDB doesn't maintain counts of keys within in a range by default), and it seemed to be of limited value. I thought we had gotten to the point where the stock advice on paging was to use `start_key` / `end_key` and `limit`. And then I just went ahead and dropped `total_rows` to avoid the extra lookup and for parity with `_changes`.

If there's an argument to keep `total_rows` that's a pretty cheap thing to include.

Reply:
In fact, the current implementation appears to include `total_rows`. It also includes an `offset` field set to `null`